### PR TITLE
Refactor testing

### PR DESCRIPTION
**Pull request by nedtwigg**
As mentioned in #190, our testing has gotten a little messy.  Most of the test resources we use are copy-pasted multiple places.  The originals live in `testlib/src/test/resources` or `lib-extra/src/test/resources`, and they get copy-pasted into `plugin-gradle` and `plugin-maven` for integration tests.

We've also got some weirdness in `ResourceHarness` - it sometimes adds newlines.  This PR is a breaking change to `testlib`, and a cleanup of this jumble.

In this first commit, all of the test resources have been combined into  `testlib/src/test/resources`, and the smallest-possible cleanup of the tests themselves to permit this.

### CODE REVIEW

**Comment by nedtwigg** [comment]
The middle commit (993390d9fae492c1c4d9db63f2fae8c7794a8b1a) is the important one - it changes all our integration tests to an (imo) easier-to-read and maintain style.

**Comment by nedtwigg** [comment]
@lutovich - this affects your recent PR, but since you haven't accepted your committer invite yet I can't add you as reviewer.  This look okay to you?

File: plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GroovyGradleExtensionTest.java [hunk 1]
```diff
@@ -47,18 +46,14 @@ private void testTarget(boolean useDefaultTarget) throws IOException {
 				"        licenseHeader('" + HEADER + "', 'plugins')",
 				"    }",
 				"}");
+		setFile("build.gradle").toContent(buildContent);
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GroovyGradleExtensionTest.java]: The `buildContent` string is used in the assertions below.

[the remark this answers is not in the document]
Comment (jbduncan) [reply plugin-gradle/src/test/java/com/diffplug/gradle/spotless/GroovyGradleExtensionTest.java]: Ah okay, in that case feel free to completely ignore my earlier comment! :stuck_out_tongue_winking_eye:

File: plugin-gradle/src/test/java/com/diffplug/gradle/spotless/JavaDefaultTargetTest.java [hunk 2]
```diff
@@ -33,27 +32,17 @@ public void integration() throws IOException {
 				"",
 				"spotless {",
 				"    java {",
-				"        googleJavaFormat()",
+				"        googleJavaFormat('1.2')",
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply plugin-gradle/src/test/java/com/diffplug/gradle/spotless/JavaDefaultTargetTest.java]: There's a tradeoff: does every default version bump to a FormatterStep require changing all integration tests?  If so, it's more work to bump versions.  If not, there's some tiny risk of a version bump causing a problem that doesn't get caught until people use it - I can't think of any issues we've had to date where a unit test on the FormatterStep passed but it failed in integration.

If you free time and see some entropy, I'm all for applying one to the other :)  But I think keeping Spotless easy to contribute to is more important than keeping its entropy low, so I'm not in favor of a policy like "unit tests must always use latest version" or something like that.  People mostly contribute PR's when they have an itch to scratch - so long as their itch doesn't create unbearable itches for us, I'd like to merge them in with a minimum of guidelines.

File: testlib/src/main/java/com/diffplug/spotless/ResourceHarness.java [hunk 3]
```diff
@@ -194,4 +148,74 @@ protected void assertOnResources(FormatterFunc step, String unformattedPath, Str
...
+		public void hasContent(String expected) {
+			hasContent(expected, StandardCharsets.UTF_8);
+		}
+
+		public void hasContent(String expected, Charset charset) {
+			Assertions.assertThat(file).usingCharset(charset).hasContent(expected);
+		}
+
+		public void hasLines(String... lines) {
+			hasContent(Arrays.stream(lines).collect(Collectors.joining("\n")));
```
[the remark this answers is not in the document]
Comment (nedtwigg) [reply testlib/src/main/java/com/diffplug/spotless/ResourceHarness.java]: You're a living javadoc for JavaSE8!  Good call.  I'll fixup with this and merge.

File: plugin-gradle/src/test/java/com/diffplug/gradle/spotless/JavaDefaultTargetTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbduncan) [reply plugin-gradle/src/test/java/com/diffplug/gradle/spotless/JavaDefaultTargetTest.java]: Sounds fair! I've just remembered actually that I was looking into updating versions from 1.x to 1.5 on my local clone of Spotless recently, so I'm happy to address this and raise my own PR for it in my own time. :)

**Comment by jbduncan** [comment]
https://github.com/diffplug/spotless/pull/203/commits/86a13c361053fd4153a2ac2d04f16d1f339d6050 looks good to me! :+1:

File: testlib/src/main/java/com/diffplug/spotless/ResourceHarness.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbduncan) [reply testlib/src/main/java/com/diffplug/spotless/ResourceHarness.java]: Haha! I'll take that as a compliment. :wink:

**Pull request merged**